In [1]:
import pandas as pd

## 1. Data Import & Preview

In [2]:
order_list = pd.read_csv('./data/order_list.csv')
station_list = pd.read_csv('./data/station_list.csv')
user_list = pd.read_csv('./data/user_list.csv')

C:\Users\happy\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\IPython\core\interactiveshell.py:3251: DtypeWarning: Columns (14) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [3]:
order_list.head()

이용자ID                        호출일시                                     호출장소  \
0  11506  0054-05-16 16:38:12.000000   (37.49335710216712,126.49071292910158)   
1      1  2021-02-15 00:00:24.000000  (37.508301689901344,126.51395666732049)   
2      7  2021-02-15 00:00:35.000000                 (37.5085858,126.5137479)   
3      1  2021-02-15 00:00:46.000000  (37.508370908943476,126.51402737025778)   
4   7901  2021-02-15 00:00:47.000000                 (37.5086122,126.5137468)   

  탑승일시  탑승정류장 하차일시  하차정류장  어른(명)  청소년(명)  어린이(명)  이동거리(m)  요금(원)   배차상태  \
0  NaN  10109  NaN  12007      1       0       0      870   1800  탑승취소    
1  NaN  10258  NaN  10093      1       0       0        0      0   배차불가   
2  NaN  10258  NaN  12005      1       0       0        0      0   배차불가   
3  NaN  10258  NaN  10093      1       0       0        0      0   배차불가   
4  NaN  10258  NaN  10196      1       0       0        0      0   배차불가   

  배차상태상세                      배차성공일시                        취소일시  
0   개인사정  2021-07-30 16:38:19.000000  2021-07-30 16:38:39.000000  
1    NaN                         NaN  2021-02-15 00:00:25.000000  
2    NaN                         NaN  2021-02-15 00:00:35.000000  
3    NaN                         NaN  2021-02-15 00:00:46.000000  
4    NaN                         NaN  2021-02-15 00:00:47.000000

In [4]:
# 319 stations

print(station_list.shape[0])
station_list.head()

319


id     name                  input_date  is_using  \
0  10011    힐스테이트  2021-01-04 14:50:54.000000      True   
1  10001    허브주유소  2021-01-04 14:50:54.000000      True   
2  10921  햇내로공원입구  2021-01-04 14:50:54.000000      True   
3  10920  햇내로공원입구  2021-01-04 14:50:54.000000      True   
4  10027   해맞이공원앞  2021-01-04 14:50:54.000000      True   

                          xy  st_no(공공정류소ID)  
0    (126.5642385,37.489032)         35458.0  
1   (126.5298091,37.5049674)         35446.0  
2       (126.50107,37.49577)         35784.0  
3        (126.50135,37.4967)         35777.0  
4  (126.5318204,37.47709801)         35616.0

In [5]:
# 11,577 distinct users

print(user_list.shape[0])
user_list.head()

11577


no               register_date
0   1  2021-02-15 00:00:24.000000
1   2  2021-02-15 00:04:24.000000
2   3  2021-02-02 17:24:00.000000
3   4  2021-02-17 15:36:51.000000
4   5  2021-02-04 10:58:00.000000

## 2. Data Cleaning

### 2-(1) Datatype Conversion

In [6]:
order_list.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3660646 entries, 0 to 3660645
Data columns (total 16 columns):
 #   Column   Dtype 
---  ------   ----- 
 0   이용자ID    int64 
 1   호출일시     object
 2   호출장소     object
 3   탑승일시     object
 4   탑승정류장    int64 
 5   하차일시     object
 6   하차정류장    int64 
 7   어른(명)    int64 
 8   청소년(명)   int64 
 9   어린이(명)   int64 
 10  이동거리(m)  int64 
 11  요금(원)    int64 
 12  배차상태     object
 13  배차상태상세   object
 14  배차성공일시   object
 15  취소일시     object
dtypes: int64(8), object(8)
memory usage: 446.9+ MB


In [3]:
# '호출일시' datetime 이상치 제거
order_list = order_list.iloc[1:-2]


In [4]:
# Convert Datatype for columns : 호출일시, 탑승일시, 하차일시, 배차성공일시, 취소일시 

def convert(x):
    x = x.str.split('.').str[0].astype('datetime64')
    return x

order_list['call_time'] = convert(order_list['호출일시'])
order_list['board_time'] = convert(order_list['탑승일시'])
order_list['alight_time'] = convert(order_list['하차일시'])
order_list['suc_time'] = convert(order_list['배차성공일시'])
order_list['cancel_time'] = convert(order_list['취소일시'])

In [5]:
order_list.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3660643 entries, 1 to 3660643
Data columns (total 21 columns):
 #   Column       Dtype         
---  ------       -----         
 0   이용자ID        int64         
 1   호출일시         object        
 2   호출장소         object        
 3   탑승일시         object        
 4   탑승정류장        int64         
 5   하차일시         object        
 6   하차정류장        int64         
 7   어른(명)        int64         
 8   청소년(명)       int64         
 9   어린이(명)       int64         
 10  이동거리(m)      int64         
 11  요금(원)        int64         
 12  배차상태         object        
 13  배차상태상세       object        
 14  배차성공일시       object        
 15  취소일시         object        
 16  call_time    datetime64[ns]
 17  board_time   datetime64[ns]
 18  alight_time  datetime64[ns]
 19  suc_time     datetime64[ns]
 20  cancel_time  datetime64[ns]
dtypes: datetime64[ns](5), int64(8), object(8)
memory usage: 586.5+ MB


In [5]:
order_list = order_list.drop(['호출일시', '탑승일시', '하차일시', '배차성공일시', '취소일시'], axis = 1)

In [6]:
# Split latitude and longitude for column '호출장소'

order_list[["x", "y"]] = order_list["호출장소"].str.strip(r"()").str.split(",", expand = True).astype(float)

order_list = order_list.drop(['호출장소'], axis = 1)

In [8]:
order_list.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3660643 entries, 1 to 3660643
Data columns (total 17 columns):
 #   Column       Dtype         
---  ------       -----         
 0   이용자ID        int64         
 1   탑승정류장        int64         
 2   하차정류장        int64         
 3   어른(명)        int64         
 4   청소년(명)       int64         
 5   어린이(명)       int64         
 6   이동거리(m)      int64         
 7   요금(원)        int64         
 8   배차상태         object        
 9   배차상태상세       object        
 10  call_time    datetime64[ns]
 11  board_time   datetime64[ns]
 12  alight_time  datetime64[ns]
 13  suc_time     datetime64[ns]
 14  cancel_time  datetime64[ns]
 15  x            float64       
 16  y            float64       
dtypes: datetime64[ns](5), float64(2), int64(8), object(2)
memory usage: 474.8+ MB


### 2-(2) Rename columns / Unit

In [7]:
order_list = order_list.rename(columns = {"이용자ID":"ID", "탑승정류장":"board_station", "하차정류장":"alight_station", "어른(명)":"adult", "청소년(명)":"teen", "어린이(명)":"child", 
                                "이동거리(m)":"distance", "요금(원)":"fare", "배차상태":"suc_status", "배차상태상세":"suc_status_d"})

In [8]:
# change unit of "distance" (m to km)

order_list['distance'] = order_list['distance'] / 1000

### 2-(3) Data 입력사항 수정 및 '배차상태' 정리

In [11]:
order_list['suc_status'].value_counts()

배차불가       3323767
시뮬레이션취소     158462
하차완료        133829
탑승취소         39235
시스템취소         3450
배차불가          1899
하차완료             1
Name: suc_status, dtype: int64

In [12]:
order_list['suc_status_d'].value_counts()

승차시간 부족     14520
승차시간 지연      7146
다른교통수단이용     6974
개인사정         6468
미탑승          3452
결제승인실패       1971
결제인증실패       1191
중복된 이용신청     1118
운행가능차량없음      781
중복신청취소        584
시스템오류         226
관리자 취소        144
탑승권생성실패        10
Name: suc_status_d, dtype: int64

현재 suc_status_d가 '운행가능차량없음' 또는 '중복된 이용신청'이라고 입력되어 있는 행의 suc_status가 '배차불가'로 입력되어 있지 않아 수정

In [9]:
order_list.loc[(order_list['suc_status_d']=='운행가능차량없음')|(order_list['suc_status_d']=='중복된 이용신청'),'suc_status'] = '배차불가'

In [10]:
# 수정 결과 확인 

order_list.loc[(order_list['suc_status']=='배차불가')&(order_list['suc_status_d']=='운행가능차량없음')]
order_list.loc[(order_list['suc_status']=='배차불가')&(order_list['suc_status_d']=='중복된 이용신청')]

ID  board_station  alight_station  adult  teen  child  distance  \
67012    10738          10090           10024      1     0      0       0.0   
67013    10738          10090           10024      1     0      0       0.0   
67014    10738          10090           10024      1     0      0       0.0   
67015    10738          10090           10024      1     0      0       0.0   
67016    10738          10090           10024      1     0      0       0.0   
...        ...            ...             ...    ...   ...    ...       ...   
3006117   5726          10031           20027      1     0      0       0.0   
3008345  11376          12005           21366      1     0      0       0.0   
3011763    969          10155           10035      0     1      0       0.0   
3497802   6345          10015           10193      0     0      1       0.0   
3599118   8858          10116           10910      0     2      0       0.0   

         fare suc_status suc_status_d           call_time board_time  \
67012       0       배차불가     중복된 이용신청 2021-02-28 12:34:42        NaT   
67013       0       배차불가     중복된 이용신청 2021-02-28 12:34:53        NaT   
67014       0       배차불가     중복된 이용신청 2021-02-28 12:35:08        NaT   
67015       0       배차불가     중복된 이용신청 2021-02-28 12:36:00        NaT   
67016       0       배차불가     중복된 이용신청 2021-02-28 12:36:13        NaT   
...       ...        ...          ...                 ...        ...   
3006117     0       배차불가     중복된 이용신청 2022-04-25 19:03:20        NaT   
3008345     0       배차불가     중복된 이용신청 2022-04-26 05:44:10        NaT   
3011763     0       배차불가     중복된 이용신청 2022-04-26 11:07:14        NaT   
3497802     0       배차불가     중복된 이용신청 2022-06-17 15:21:43        NaT   
3599118     0       배차불가     중복된 이용신청 2022-06-27 16:36:29        NaT   

        alight_time suc_time         cancel_time          x           y  
67012           NaT      NaT 2021-02-28 12:34:43  37.483058  126.507528  
67013           NaT      NaT 2021-02-28 12:34:54  37.483045  126.507538  
67014           NaT      NaT 2021-02-28 12:35:09  37.483059  126.507570  
67015           NaT      NaT 2021-02-28 12:36:01  37.483183  126.507730  
67016           NaT      NaT 2021-02-28 12:36:14  37.483142  126.507606  
...             ...      ...                 ...        ...         ...  
3006117         NaT      NaT 2022-04-25 19:03:21  37.488303  126.555161  
3008345         NaT      NaT 2022-04-26 05:44:11  37.490666  126.496574  
3011763         NaT      NaT 2022-04-26 11:07:14  37.485221  126.501534  
3497802         NaT      NaT 2022-06-17 15:21:43  37.491743  126.553823  
3599118         NaT      NaT 2022-06-27 16:36:43  37.497895  126.486392  

[1118 rows x 17 columns]

현재 suc_status_d가 '결제승인실패','결제인증실패','시스템오류','탑승권생성실패' 라고 입력되어 있는 행의 suc_status가 '시스템취소'로 입력되어 있지 않아 수정

In [11]:

order_list.loc[(order_list['suc_status_d']=='결제승인실패')|(order_list['suc_status_d']=='결제인증실패')|
          (order_list['suc_status_d']=='시스템오류')|(order_list['suc_status_d']=='탑승권생성실패'),'suc_status'] = '시스템취소'
          
order_list.loc[(order_list['suc_status']=='시스템취소')]

ID  board_station  alight_station  adult  teen  child  distance  \
10700    11081          10120           10057      1     0      0      1.78   
10702    11156          10035           10051      1     0      0      5.99   
10703    11084          10172           10012      0     1      0      5.98   
10711      146          10152           10165      1     0      0      9.71   
10715      146          10152           10165      1     0      0      9.71   
...        ...            ...             ...    ...   ...    ...       ...   
3654170    698          10127           10910      1     0      0      0.78   
3657155   5426          10120           10148      0     1      0      1.38   
3660346   9834          10120           10161      0     1      0      6.90   
3660562   7196          10911           21364      1     0      0      4.91   
3660567   7196          10911           21364      2     0      0      4.91   

         fare suc_status suc_status_d           call_time board_time  \
10700    1800      시스템취소       결제인증실패 2021-02-17 14:32:29        NaT   
10702    1800      시스템취소       결제인증실패 2021-02-17 14:32:34        NaT   
10703    1200      시스템취소       결제인증실패 2021-02-17 14:32:43        NaT   
10711    2000      시스템취소       결제인증실패 2021-02-17 14:33:14        NaT   
10715    2000      시스템취소       결제인증실패 2021-02-17 14:33:36        NaT   
...       ...        ...          ...                 ...        ...   
3654170  1800      시스템취소       결제승인실패 2022-07-02 13:46:00        NaT   
3657155  1200      시스템취소       결제승인실패 2022-07-02 17:37:10        NaT   
3660346  1200      시스템취소       결제승인실패 2022-07-02 22:00:42        NaT   
3660562  1800      시스템취소       결제승인실패 2022-07-02 22:30:19        NaT   
3660567  3600      시스템취소       결제승인실패 2022-07-02 22:31:19        NaT   

        alight_time            suc_time         cancel_time          x  \
10700           NaT                 NaT 2021-02-17 14:32:36  37.493371   
10702           NaT                 NaT 2021-02-17 14:32:40  37.492069   
10703           NaT                 NaT 2021-02-17 14:32:47  37.518023   
10711           NaT                 NaT 2021-02-17 14:33:23  37.495410   
10715           NaT                 NaT 2021-02-17 14:33:42  37.495410   
...             ...                 ...                 ...        ...   
3654170         NaT 2022-07-02 13:46:06 2022-07-02 13:46:07  37.494044   
3657155         NaT 2022-07-02 17:37:18 2022-07-02 17:37:20  37.493473   
3660346         NaT 2022-07-02 22:00:55 2022-07-02 22:00:59  37.494271   
3660562         NaT 2022-07-02 22:30:28 2022-07-02 22:30:30  37.489144   
3660567         NaT 2022-07-02 22:31:24 2022-07-02 22:31:29  37.489144   

                  y  
10700    126.490772  
10702    126.561908  
10703    126.538411  
10711    126.488948  
10715    126.488948  
...             ...  
3654170  126.566541  
3657155  126.490613  
3660346  126.488623  
3660562  126.559191  
3660567  126.559191  

[3398 rows x 17 columns]

In [12]:
order_list.loc[(order_list['suc_status']=='탑승취소 '),'suc_status'] = '탑승취소'

order_list.loc[(order_list['suc_status']=='시스템취소 '),'suc_status'] = '시스템취소'

order_list.iloc[164426,8] = '하차완료'

최종 수정 결과 확인

In [24]:
print(order_list['suc_status'].value_counts())
print(order_list['suc_status_d'].value_counts())

배차불가       3325666
시뮬레이션취소     158462
하차완료        133830
탑승취소         39235
시스템취소         3450
Name: suc_status, dtype: int64
승차시간 부족     14520
승차시간 지연      7146
다른교통수단이용     6974
개인사정         6468
미탑승          3452
결제승인실패       1971
결제인증실패       1191
중복된 이용신청     1118
운행가능차량없음      781
중복신청취소        584
시스템오류         226
관리자 취소        144
탑승권생성실패        10
Name: suc_status_d, dtype: int64


## 3. 중복 기록된 호출이력 삭제

동일 ID의 사람이 정확히 같은 시간에 같은 승/하차 정류장 & 같은 거리 & 요금 & 탑승인원으로 호출한 경우가 있어, 

해당 경우는 중복된 호출들 중 가장 마지막 호출만 남겨둔다. 

In [13]:
order_list = order_list.drop_duplicates(subset = ['ID','call_time','board_station','alight_station','distance','fare','adult','teen','child'],keep = 'last', inplace = False, ignore_index = True)

In [18]:
order_list.shape

(3657176, 17)

## 4. QGIS 이용하여 각 정류장 별 소속 법정동 위치 라벨링

In [14]:
station_adm = pd.read_csv("./data/station_list_ADM.csv")
station_adm.head()

id     name               input_date  is_using           x          y  \
0  10011    힐스테이트  2021/01/04 14:50:54.000      True  126.564239  37.489032   
1  10001    허브주유소  2021/01/04 14:50:54.000      True  126.529809  37.504967   
2  10921  햇내로공원입구  2021/01/04 14:50:54.000      True  126.501070  37.495770   
3  10920  햇내로공원입구  2021/01/04 14:50:54.000      True  126.501350  37.496700   
4  10027   해맞이공원앞  2021/01/04 14:50:54.000      True  126.531820  37.477098   

     st_no      ADM  
0  35458.0      중산동  
1  35446.0      운북동  
2  35784.0  운서동(주거)  
3  35777.0  운서동(주거)  
4  35616.0      운남동

In [18]:
station_adm['ADM'].unique()

array(['중산동', '운북동', '운서동(주거)', '운남동', '운서동(공항)', '을왕동', '남북동', '덕교동'],
      dtype=object)

In [15]:
station_adm = station_adm.drop(columns=['name', 'input_date','is_using','x','y','st_no'], axis=1)

In [16]:
order_list = order_list.merge(station_adm, how = 'left', left_on= 'board_station', right_on = 'id')
order_list = order_list.drop(columns = ['id'], axis=1).rename(columns = {"ADM":"board_adm"})

In [17]:
order_list = order_list.merge(station_adm, how = 'left', left_on= 'alight_station', right_on = 'id')
order_list = order_list.drop(columns = ['id'], axis=1).rename(columns = {"ADM":"alight_adm"})
order_list.head(3)

ID  board_station  alight_station  adult  teen  child  distance  fare  \
0   1          10258           10093      1     0      0       0.0     0   
1   7          10258           12005      1     0      0       0.0     0   
2   1          10258           10093      1     0      0       0.0     0   

  suc_status suc_status_d           call_time board_time alight_time suc_time  \
0       배차불가          NaN 2021-02-15 00:00:24        NaT         NaT      NaT   
1       배차불가          NaN 2021-02-15 00:00:35        NaT         NaT      NaT   
2       배차불가          NaN 2021-02-15 00:00:46        NaT         NaT      NaT   

          cancel_time          x           y board_adm alight_adm  
0 2021-02-15 00:00:25  37.508302  126.513957       운북동        운남동  
1 2021-02-15 00:00:35  37.508586  126.513748       운북동    운서동(주거)  
2 2021-02-15 00:00:46  37.508371  126.514027       운북동        운남동

정리된 order table 내보내기 

In [23]:
order_list.to_csv('./data/order.csv')

# 5. O/D  이력 만들기

In [18]:
order = pd.read_csv('./data/order.csv')
order = order.drop(['Unnamed: 0'], axis = 1)

C:\Users\happy\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\IPython\core\interactiveshell.py:3251: DtypeWarning: Columns (14) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [25]:
order.shape

(3657176, 19)

In [26]:
order.head(3)

ID  board_station  alight_station  adult  teen  child  distance  fare  \
0   1          10258           10093      1     0      0       0.0     0   
1   7          10258           12005      1     0      0       0.0     0   
2   1          10258           10093      1     0      0       0.0     0   

  suc_status suc_status_d            call_time board_time alight_time  \
0       배차불가          NaN  2021-02-15 00:00:24        NaN         NaN   
1       배차불가          NaN  2021-02-15 00:00:35        NaN         NaN   
2       배차불가          NaN  2021-02-15 00:00:46        NaN         NaN   

  suc_time          cancel_time          x           y board_adm alight_adm  
0      NaN  2021-02-15 00:00:25  37.508302  126.513957       운북동        운남동  
1      NaN  2021-02-15 00:00:35  37.508586  126.513748       운북동    운서동(주거)  
2      NaN  2021-02-15 00:00:46  37.508371  126.514027       운북동        운남동

현재 'order' table의 나열은 'call_time(호출시간)' 오름차순 

이를 ID별로 묶어서 call_time 오름차순으로 나열해보자.

In [19]:
order_sort = order.sort_values(['ID', 'call_time'], ignore_index=True)
order_sort.head()

ID  board_station  alight_station  adult  teen  child  distance  fare  \
0   1          10258           10093      1     0      0      0.00     0   
1   1          10258           10093      1     0      0      0.00     0   
2   1          10258           10093      1     0      0      0.00     0   
3   1          10258           10203      1     0      0      3.58  1800   
4   1          10263           10093      1     0      0      5.18  1800   

  suc_status suc_status_d            call_time board_time alight_time  \
0       배차불가          NaN  2021-02-15 00:00:24        NaN         NaN   
1       배차불가          NaN  2021-02-15 00:00:46        NaN         NaN   
2       배차불가          NaN  2021-02-15 00:00:59        NaN         NaN   
3       탑승취소      승차시간 부족  2021-02-15 00:04:17        NaN         NaN   
4       탑승취소      승차시간 부족  2021-02-15 00:19:49        NaN         NaN   

  suc_time          cancel_time          x           y board_adm alight_adm  
0      NaN  2021-02-15 00:00:25  37.508302  126.513957       운북동        운남동  
1      NaN  2021-02-15 00:00:46  37.508371  126.514027       운북동        운남동  
2      NaN  2021-02-15 00:00:59  37.508506  126.513753       운북동        운남동  
3      NaN  2021-02-15 00:06:01  37.508302  126.513956       운북동        운북동  
4      NaN  2021-02-15 00:20:12  37.508806  126.513949       운북동        운남동

In [29]:
order_byid = order[['ID']].value_counts()
order_byid

ID   
11188    26278
2362     19193
2639     17643
9438     17206
10910    14105
         ...  
4743         1
6634         1
6633         1
6632         1
9591         1
Length: 11577, dtype: int64

ID별로 나열한 order_sort table과 ID별 호출 횟수를 나열한 order_byid table 두 개를 모두 봤을 때, 한 사람이 단기에 너무 많이 호출 - 호출취소를 한 이력이 보인다. 

ID '11188' 이용자는 총 26,278번의 이력 (최대)가 존재한다. 

유저 ID #1 인 사람이 첫 10번의 호출간 어떻게 이동했는지 관찰해보자. 

In [30]:
# show how the location of the user #1 changes 

# import package 
import folium

# make a map
user_1 = folium.Map(location=[37.508, 126.513], tiles = "OpenStreetMap", zoom_start=16)

# add the location of user #1 from row 0-8 on table 'order_sort'
for i in range(0,9):
    folium.Marker(
        location = [order_sort.iloc[i]['x'], order_sort.iloc[i]['y']]
    ).add_to(user_1)

# show the map
user_1

In [28]:
# 다른 id로 추가 관찰

order_sort.loc[order_sort['ID'] == 3983].head(5)

ID  board_station  alight_station  adult  teen  child  distance  \
1145404  3983          20022           10142      1     0      0      0.00   
1145405  3983          20022           10142      1     0      0      8.90   
1145406  3983          10142           10400      1     0      0     14.01   
1145407  3983          10911           10120      1     0      0      7.97   
1145408  3983          10032           10120      1     0      0      7.88   

         fare suc_status suc_status_d            call_time  \
1145404     0       배차불가          NaN  2021-09-08 14:46:47   
1145405  1900       하차완료          NaN  2021-09-08 14:47:00   
1145406  2500       탑승취소      승차시간 부족  2021-12-27 13:21:22   
1145407  1800       탑승취소      승차시간 부족  2022-04-20 15:08:21   
1145408  1800       하차완료          NaN  2022-04-20 15:10:56   

                  board_time          alight_time             suc_time  \
1145404                  NaN                  NaN                  NaN   
1145405  2021-09-08 14:55:37  2021-09-08 15:07:42  2021-09-08 14:47:12   
1145406                  NaN                  NaN  2021-12-27 13:21:31   
1145407                  NaN                  NaN  2022-04-20 15:08:34   
1145408  2022-04-20 15:27:48  2022-04-20 15:45:20  2022-04-20 15:11:10   

                 cancel_time          x           y board_adm alight_adm  
1145404  2021-09-08 14:46:52  37.489168  126.477156   운서동(공항)        중산동  
1145405                  NaN  37.489168  126.477156   운서동(공항)        중산동  
1145406  2021-12-27 13:22:24  37.491609  126.551100       중산동    운서동(공항)  
1145407  2022-04-20 15:09:15  37.488818  126.559355       중산동    운서동(주거)  
1145408                  NaN  37.489083  126.558878       중산동    운서동(주거)

### 3-(1) '배차상태'로 실이동수요를 살펴볼 수 있을까?에 대한 고찰

In [29]:
order_sort[order_sort['board_time'].notnull()].count()[0]
order_sort[order_sort['alight_time'].notnull()].count()[0]

# '승차시각' 및 '하차시각'이 기록되어 있는 이력의 수는 133,830건

133830

In [30]:
order_sort[order_sort['board_time'].notnull()]['suc_status'].value_counts()

# '승차시각'이 기록되어 있는 모든 이력의 '배차상태'는 '하차완료'이다. 
order_sort[order_sort['alight_time'].notnull()]['suc_status'].value_counts()

# 마찬가지로,, '하차시각'이 기록되어 있는 모든 이력의 '배차상태'는 '하차완료'이다. 

하차완료    133830
Name: suc_status, dtype: int64

In [31]:
print(order_sort[order_sort['suc_time'].notnull()].count()[0])
print(order_sort[order_sort['cancel_time'].notnull()].count()[0])


169302
3527268


In [32]:
# 배차성공한 이력의 배차 상태 분류
order_sort[order_sort['suc_time'].notnull()]['suc_status'].value_counts()

하차완료       130082
탑승취소        37180
시스템취소        2037
시뮬레이션취소         3
Name: suc_status, dtype: int64

In [36]:
# 배차성공한 이력의 배차 상태 상세 분류
order_sort[order_sort['suc_time'].notnull()]['suc_status_d'].value_counts()

승차시간 부족     13282
승차시간 지연      7146
다른교통수단이용     6678
개인사정         6177
미탑승          3245
결제승인실패       1971
중복신청취소        584
관리자 취소        121
결제인증실패          8
시스템오류           4
탑승권생성실패         2
Name: suc_status_d, dtype: int64

In [37]:
# 배차성공 못한 이력의 배차 상태 분류
order_sort[order_sort['suc_time'].isnull()]['suc_status'].value_counts()

배차불가       3322398
시뮬레이션취소     158261
하차완료          3748
탑승취소          2055
시스템취소         1412
Name: suc_status, dtype: int64

In [38]:
# 배차 취소한 이력의 배차 상태 상세 분류
order_sort[order_sort['cancel_time'].notnull()]['suc_status'].value_counts()

배차불가       3322398
시뮬레이션취소     158264
탑승취소         39235
하차완료          3928
시스템취소         3443
Name: suc_status, dtype: int64

In [39]:
# 배차 취소하지 않은 이력의 배차 상태 분류
order_sort[order_sort['cancel_time'].isnull()]['suc_status'].value_counts()

하차완료     129902
시스템취소         6
Name: suc_status, dtype: int64

결론 : '배차 상태'로 실 이동수요의 묶음을 그룹화하기는 어려움, 기준의 애매함

다른 방법 : 직전행과의 'board_station' (탑승희망정류장)이 같은지, 다른지 비교하기

### 3-(2) 탑승희망 정류장의 변화 여부 관찰

In [20]:
# Trip Chain

order_sort['Trip'] = 0

# Check if the alighting station is different from the station of previous row

# Function to update the 'Trip' column based on the conditions
def update_trip(row):
    global prev_trip,prev_station
    if row['board_station'] != prev_station:
        prev_trip += 1
    else:
        pass
    prev_station = row['board_station']
    return prev_trip

# Set the initial value for prev_trip
prev_trip = 0
prev_station = order_sort.at[0, 'board_station']

# Update the 'Trip' column using the apply function
order_sort['Trip'] = order_sort.apply(update_trip, axis=1)



In [21]:
order_sort.to_csv('./data/order_sort.csv')

order_last 테이블에 trip number를 부여한 후, trip number가 같은 것 중 최종 행만 이동 실수요로 생각. >> 'order_last' 파일로 export

In [34]:
order_last = order_sort.drop_duplicates(subset=['Trip'], keep='last', ignore_index=True )

In [35]:
order_last.to_csv('./data/order_last.csv')